examples/vision/siamesenetwork.py

```diff
@@ -0,0 +1,417 @@
+        preprocess_image(positive),
+        preprocess_image(negative),
+    )
+
+
+"""
+Let's setup our data pipeline using a zipped list with an anchor, positive,
+and negative image filename as the source. The output of the pipeline
+contains the same triplet with every image loaded and preprocessed.
+"""
+
+# We need to make sure both the anchor and positive images are loaded in
+# sorted order so we can match them together.
+anchor_images = sorted(
+    [str(anchor_images_path / f) for f in os.listdir(anchor_images_path)]
+)
+
+positive_images = sorted(
+    [str(positive_images_path / f) for f in os.listdir(positive_images_path)]
+)
+
+image_count = len(anchor_images)
+
+anchor_dataset = tf.data.Dataset.from_tensor_slices(anchor_images)
+positive_dataset = tf.data.Dataset.from_tensor_slices(positive_images)
+
+# To generate the list of negative images, let's randomize the list of
+# available images and concatenate them together.
+rng = np.random.RandomState(seed=42)
+rng.shuffle(anchor_images)
+rng.shuffle(positive_images)
+
+negative_images = anchor_images + positive_images
+np.random.RandomState(seed=32).shuffle(negative_images)
+
+negative_dataset = tf.data.Dataset.from_tensor_slices(negative_images)
+negative_dataset = negative_dataset.shuffle(buffer_size=4096)
+
+dataset = tf.data.Dataset.zip((anchor_dataset, positive_dataset, negative_dataset))
+dataset = dataset.shuffle(buffer_size=1024)
+dataset = dataset.map(preprocess_triplets)
+dataset = dataset.batch(32, drop_remainder=False)
+dataset = dataset.prefetch(tf.data.AUTOTUNE)
+
+# Let's now split our dataset in train and validation.
+train_dataset = dataset.take(round(image_count * 0.8))
+val_dataset = dataset.skip(round(image_count * 0.8))
+
+
+"""
+Let's take a look at a few examples of triplets. Notice how the first two images
+look alike while the third is always different.
+"""
+
+
+def visualize(anchor, positive, negative):
+    """
+    Visualizes a few triplets from the supplied batches.
+    """
+
+    def show(ax, image):
+        ax.imshow(image)
+        ax.get_xaxis().set_visible(False)
+        ax.get_yaxis().set_visible(False)
+
+    fig = plt.figure(figsize=(9, 9))
+
+    axs = fig.subplots(3, 3)
+    for i in range(3):
+        show(axs[i, 0], anchor[i])
+        show(axs[i, 1], positive[i])
+        show(axs[i, 2], negative[i])
+
+
+visualize(*list(dataset.take(1).as_numpy_iterator())[0])
+
+"""
+# Setting up the Embedding generator model
+
+Our Siamese Network will generate embeddings for each one of the images of the
+triplet. To do this, we will use a pre-trained ResNet50 model on ImageNet and
+connect a few `Dense` layers to it so we have space to learn to separate these
+embeddings.
+
+We will freeze the weights of all the layers of the model up until
+`conv5_block1_out`. This is important so we don't mess with the weights that
+the model already learned. We are going to leave the bottom few layers open so
+we can fine tune those weights during training.
```

Comment:
Suggestion:

```suggestion
We will freeze the weights of all the layers of the model up until `conv5_block1_out`. 
This is important to avoid affecting the weights that the model has already learned. 
We are going to leave the bottom few layers open, so that we can fine-tune their weights
during training.
```